### PR TITLE
[API] candidate creation

### PR DESCRIPTION
**Pull request by xlecours**
This pull request make the EDC property optionnal has specified in the documentation. It also take the Project name into account if specified.

See: https://github.com/aces/Loris/blob/17.1-dev/docs/API/LorisRESTAPI.md#30-candidate-api
redmine: 12583

Additionally, it remove the else(s) keywords after a the safeExit() calls

### CODE REVIEW

File: htdocs/api/v0.0.2/Candidates.php [hunk 1]
```diff
@@ -106,13 +106,14 @@ public function handlePOST()
         if (isset($this->RequestData['Candidate'])) {
             $data = $this->RequestData;
             if ($data === null) {
+                error_log("Can't parse data");
                 $this->header("HTTP/1.1 400 Bad Request");
                 $this->safeExit(0);
             }
 
             $this->verifyField($data, 'Gender', ['Male', 'Female']);
-            $this->verifyField($data, 'EDC', 'YYYY-MM-DD');
```
Comment (jstirling91) [inline htdocs/api/v0.0.2/Candidates.php]: Should this not check use if useEDC is enabled, verifying if so?

File: htdocs/api/v0.0.2/Candidates.php [hunk 2]
```diff
@@ -106,13 +106,14 @@ public function handlePOST()
         if (isset($this->RequestData['Candidate'])) {
             $data = $this->RequestData;
             if ($data === null) {
+                error_log("Can't parse data");
```
Comment (jstirling91) [inline htdocs/api/v0.0.2/Candidates.php]: Is this for debugging?

File: htdocs/api/v0.0.2/Candidates.php (hunk 1 above)
Comment (xlecours) [inline htdocs/api/v0.0.2/Candidates.php]: @jstirling91 Candidate class should in the createNew function

File: htdocs/api/v0.0.2/Candidates.php (hunk 2 above)
Comment (xlecours) [inline htdocs/api/v0.0.2/Candidates.php]: I'll change this for HTTP header 500

File: htdocs/api/v0.0.2/Candidates.php [hunk 3]
```diff
@@ -121,17 +121,29 @@ public function handlePOST()
                     $data['Candidate']['Gender'],
                     $data['Candidate']['PSCID']
                 );
+
+                if (isset($data['Candidate']['Project'])) {
+                    $projectName = htmlspecialchars($data['Candidate']['Project']);
+                    $project     = \Project::singleton($projectName);
+                    if (!empty($project)) {
```
Comment (jstirling91) [inline htdocs/api/v0.0.2/Candidates.php]: should an error be thrown if project doesn't exist?

Comment (xlecours) [inline htdocs/api/v0.0.2/Candidates.php]: It is thrown in \Project::singleton

**Comment by kongtiaowang** [comment]
screen shot 2017-06-14 at 1 52 11 pm

File: htdocs/api/v0.0.2/Candidates.php [hunk 4]
```diff
@@ -121,17 +121,29 @@ public function handlePOST()
                     $data['Candidate']['Gender'],
                     $data['Candidate']['PSCID']
                 );
+
+                if (isset($data['Candidate']['Project'])) {
+                    $projectName = htmlspecialchars($data['Candidate']['Project']);
```
Comment (driusan) [inline htdocs/api/v0.0.2/Candidates.php]: Why is this being HTML escaped? This is dealing with json, not HTML. a "<" or "&" is perfectly valid within the context of JSON.

**Comment by xlecours** [comment]
Removing `Pass manual testing` because I changed the PR.
Bypassing safeExit function was allowing the creation of candidates without fields at all. This is a application layer validation only.

File: SQL/0000-00-00-schema.sql [hunk 5]
```diff
@@ -206,6 +206,7 @@ CREATE TABLE `users` (
   `State` varchar(255) default NULL,
   `Zip_code` varchar(255) default NULL,
   `Country` varchar(255) default NULL,
+  `Phone` varchar(15) default NULL,
```
Comment (driusan) [inline SQL/0000-00-00-schema.sql]: Is this line supposed to be changed? It looks like the result of an errant rebase.

**Comment by driusan** [comment]
This was moved out of Needs Review without a review being done..

**Comment by kongtiaowang** [comment]
screen shot 2017-07-12 at 11 17 57 am

**Comment by driusan** [comment]
For anyone reviewing this, it's easier to understand the changes with https://github.com/aces/Loris/pull/2863/files?w=1 ... the indentation level of a code block changed, so it looks like a bigger change than it should be without the `w=1` parameter..

**Pull request merged**